### PR TITLE
Use rmm allocator in CAGRA prune

### PR DESCRIPTION
**Pull request by enp1s0**
This PR modifies the CAGRA prune function to use rmm allocator instead of a custom memory allocator. And also removes the multi-GPU functionality that is not suitable for the [RAFT policy](https://docs.rapids.ai/api/raft/stable/developer_guide/#multi-gpu).

rel: #1461

### CODE REVIEW

**Comment by enp1s0** [comment]
@tfeher Can you review the code and let me know if there are any problems?

File: cpp/include/raft/neighbors/detail/cagra/graph_core.cuh [hunk 1]
```diff
@@ -160,7 +140,7 @@ __global__ void kern_sort(
     uint32_t next_mask = mask << 1;
```
Comment (tfeher) [inline cpp/include/raft/neighbors/detail/cagra/graph_core.cuh:140]: I am wondering if RAFT's [bitonic sort](https://github.com/rapidsai/raft/blob/cc4a76ba19b4b01ca5cbf454365f5f48f4bf313c/cpp/include/raft/util/bitonic_sort.cuh#L43) method can be utilized here. (If yes, then it should probably be a separate PR).

File: cpp/include/raft/neighbors/detail/cagra/graph_core.cuh [hunk 2]
```diff
@@ -538,309 +386,248 @@ void prune(raft::device_resources const& res,
...
+    //
+    // Prune kNN graph
+    //
+    auto d_input_graph = raft::make_device_matrix<IdxT, IdxT>(res, graph_size, input_graph_degree);
+
+    auto detour_count = raft::make_host_matrix<uint8_t, IdxT>(graph_size, input_graph_degree);
+    auto d_detour_count =
+      raft::make_device_matrix<uint8_t, IdxT>(res, graph_size, input_graph_degree);
+    RAFT_CUDA_TRY(cudaMemset(
+      d_detour_count.data_handle(), 0xff, graph_size * input_graph_degree * sizeof(uint8_t)));
```
Comment (tfeher) [inline cpp/include/raft/neighbors/detail/cagra/graph_core.cuh:405]: While practically it does not really matter, it is more consistent if we use `cudaMemsetAsync`
```suggestion
    RAFT_CUDA_TRY(cudaMemsetAsync(
      d_detour_count.data_handle(), 0xff, graph_size * input_graph_degree * sizeof(uint8_t), res.get_stream()));
```

File: cpp/include/raft/neighbors/detail/cagra/graph_core.cuh [hunk 3]
```diff
@@ -538,309 +386,248 @@ void prune(raft::device_resources const& res,
...
+
+    auto d_num_no_detour_edges = raft::make_device_vector<uint32_t, IdxT>(res, graph_size);
+    RAFT_CUDA_TRY(
+      cudaMemset(d_num_no_detour_edges.data_handle(), 0x00, graph_size * sizeof(uint32_t)));
```
Comment (tfeher) [inline cpp/include/raft/neighbors/detail/cagra/graph_core.cuh:409]:
```suggestion
      cudaMemsetAsync(d_num_no_detour_edges.data_handle(), 0x00, graph_size * sizeof(uint32_t)), res.get_stream());
```

File: cpp/include/raft/neighbors/detail/cagra/graph_core.cuh [hunk 4]
```diff
@@ -538,309 +386,248 @@ void prune(raft::device_resources const& res,
...
-  RAFT_CUDA_TRY(cudaDeviceSynchronize());
-  RAFT_CUDA_TRY(cudaSetDevice(0));
-  RAFT_LOG_DEBUG("\n");
+    const uint32_t batch_size =
+      std::min(static_cast<uint32_t>(graph_size), static_cast<uint32_t>(256 * 1024));
+    const uint32_t num_batch = (graph_size + batch_size - 1) / batch_size;
+    const dim3 threads_prune(32, 1, 1);
+    const dim3 blocks_prune(batch_size, 1, 1);
+
+    RAFT_CUDA_TRY(cudaMemset(dev_stats.data_handle(), 0, sizeof(uint64_t) * 2));
```
Comment (tfeher) [inline cpp/include/raft/neighbors/detail/cagra/graph_core.cuh:461]:
```suggestion
    RAFT_CUDA_TRY(cudaMemsetAsync(dev_stats.data_handle(), 0, sizeof(uint64_t) * 2), res.get_stream());
```

File: cpp/include/raft/neighbors/detail/cagra/graph_core.cuh [hunk 5]
```diff
@@ -538,309 +386,248 @@ void prune(raft::device_resources const& res,
...
-                                                    i_gpu,
-                                                    batch_size,
-                                                    i_batch,
-                                                    d_detour_count[i_gpu],
-                                                    d_num_no_detour_edges[i_gpu],
-                                                    dev_stats[i_gpu]);
+    if (input_graph_degree <= MAX_DEGREE) {
+      kernel_prune = kern_prune<MAX_DEGREE, IdxT>;
+    } else {
+      fprintf(stderr,
```
Comment (tfeher) [inline cpp/include/raft/neighbors/detail/cagra/graph_core.cuh:448]:
```suggestion
      RAFT_LOG_ERROR(
```

File: cpp/include/raft/neighbors/detail/cagra/graph_core.cuh [hunk 6]
```diff
@@ -538,309 +386,248 @@ void prune(raft::device_resources const& res,
...
+        output_graph_degree,
+        batch_size,
+        i_batch,
+        d_detour_count.data_handle(),
+        d_num_no_detour_edges.data_handle(),
+        dev_stats.data_handle());
+      res.sync_stream();
+      fprintf(stderr,
+              "# Pruning kNN Graph on GPUs (%.1lf %%)\r",
+              (double)std::min<IdxT>((i_batch + 1) * batch_size, graph_size) / graph_size * 100);
```
Comment (tfeher) [inline cpp/include/raft/neighbors/detail/cagra/graph_core.cuh:477]:
```suggestion
      RAFT_LOG_DEBUG(
              "# Pruning kNN Graph on GPUs (%.1lf %%)\r",
              (double)std::min<IdxT>((i_batch + 1) * batch_size, graph_size) / graph_size * 100);
```
Alternatively, we could use RAFT_LOG_TRACE as well.

File: cpp/include/raft/neighbors/detail/cagra/graph_core.cuh [hunk 7]
```diff
@@ -538,309 +386,248 @@ void prune(raft::device_resources const& res,
...
-
-  uint32_t* dest_nodes;     // [graph_size]
-  dest_nodes = (uint32_t*)malloc(sizeof(uint32_t) * graph_size);
-  uint32_t** d_dest_nodes;  // [num_gpus][graph_size]
-  d_dest_nodes = (uint32_t**)malloc(sizeof(uint32_t*) * num_gpus);
-  for (int i_gpu = 0; i_gpu < num_gpus; i_gpu++) {
-    RAFT_CUDA_TRY(cudaSetDevice(i_gpu));
-    RAFT_CUDA_TRY(cudaMalloc(&(d_dest_nodes[i_gpu]), sizeof(uint32_t) * graph_size));
+    const double time_prune_end = cur_time();
+    fprintf(stderr,
```
Comment (tfeher) [inline cpp/include/raft/neighbors/detail/cagra/graph_core.cuh:512]:
```suggestion
   RAFT_LOG_DEBUG(
```

File: cpp/include/raft/neighbors/detail/cagra/graph_core.cuh [hunk 8]
```diff
@@ -538,309 +386,248 @@ void prune(raft::device_resources const& res,
...
+
+  {
+    //
+    // Make reverse graph
+    //
+    const double time_make_start = cur_time();
+
+    auto d_rev_graph = raft::make_device_matrix<IdxT, IdxT>(res, graph_size, output_graph_degree);
+    RAFT_CUDA_TRY(
+      cudaMemset(d_rev_graph.data_handle(), 0xff, graph_size * output_graph_degree * sizeof(IdxT)));
```
Comment (tfeher) [inline cpp/include/raft/neighbors/detail/cagra/graph_core.cuh:533]:
```suggestion
      cudaMemsetAsync(d_rev_graph.data_handle(), 0xff, graph_size * output_graph_degree * sizeof(IdxT)), res.get_stream());
```

File: cpp/include/raft/neighbors/detail/cagra/graph_core.cuh [hunk 9]
```diff
@@ -538,309 +386,248 @@ void prune(raft::device_resources const& res,
...
+
+    auto d_rev_graph_count = raft::make_device_vector<uint32_t, IdxT>(res, graph_size);
+    RAFT_CUDA_TRY(cudaMemset(d_rev_graph_count.data_handle(), 0x00, graph_size * sizeof(uint32_t)));
```
Comment (tfeher) [inline cpp/include/raft/neighbors/detail/cagra/graph_core.cuh:536]:
```suggestion
    RAFT_CUDA_TRY(cudaMemsetAsync(d_rev_graph_count.data_handle(), 0x00, graph_size * sizeof(uint32_t)), res.get_stream());
```

File: cpp/include/raft/neighbors/detail/cagra/graph_core.cuh [hunk 10]
```diff
@@ -538,309 +386,248 @@ void prune(raft::device_resources const& res,
...
-      if (pos == output_graph_degree) { num_replaced_edges += 1; }
+    for (uint64_t i = 0; i < graph_size; i++) {
+      for (uint64_t k = 0; k < output_graph_degree; k++) {
+        const uint64_t j   = pruned_graph.data_handle()[k + (output_graph_degree * i)];
+        const uint64_t pos = pos_in_array<uint32_t>(
+          j, output_graph_ptr + (output_graph_degree * i), output_graph_degree);
+        if (pos == output_graph_degree) { num_replaced_edges += 1; }
+      }
     }
+    fprintf(stderr,
```
Comment (tfeher) [inline cpp/include/raft/neighbors/detail/cagra/graph_core.cuh:626]:
```suggestion
    RAFT_LOG_DEBUG(
```

**Review by tfeher — CHANGES_REQUESTED** [review]
Thanks @enp1s0 for the PR! It looks good overall, I have just a few smaller comments.

File: cpp/include/raft/neighbors/detail/cagra/graph_core.cuh (hunk 1 above)
Comment (enp1s0) [reply cpp/include/raft/neighbors/detail/cagra/graph_core.cuh]: The bitonic sort method can be partially used in the `kern_sort` function, specifically the following part.
https://github.com/rapidsai/raft/blob/1d1c5234f6d6b43f14f51a269c186f4b0ee1233a/cpp/include/raft/neighbors/detail/cagra/graph_core.cuh#L180-L186
Is it better to use the RAFT method instead of the current method?
(I think the current one is also fine because this part will not be a performance bottleneck, and it is consistent with the rest of the function.)

**Comment by enp1s0** [comment]
Thanks @tfeher for your review! I have modified the issues you have pointed out.

File: cpp/include/raft/neighbors/detail/cagra/graph_core.cuh (hunk 1 above)
Comment (tfeher) [reply cpp/include/raft/neighbors/detail/cagra/graph_core.cuh]: For this PR it is fine to keep the current method. We shall return to this question later. The goal is improve maintainability of the codebase, by reusing existing methods. But if the existing method fits our purpose only partially, then we should think over what is the cleanest solution.

**Review by tfeher — APPROVED** [review]
Thanks @enp1s0 for addressing the issues! The PR looks good to me.
